File: 0413-arithmetic-slices/0413-arithmetic-slices.js
/**
 * @param {number[]} nums
 * @return {number}
 */
let arr;
let dp = new Array(5005);
var numberOfArithmeticSlices = function(nums) {
    arr = nums;
    dp.fill(-1)
    let ans = 0;
    for(let i = 0; i<nums.length; i++){
         ans += f(i);
    }
    return ans;
};

function f(i){
    if(i == 0 || i == 1) return 0;

    if(dp[i] != -1)return dp[i];

    if(arr[i] - arr[i-1] != arr[i-1] - arr[i-2]){
        return dp[i] = 0;
    }else{
        return dp[i] = 1 + f(i-1);
    }
}

File: 0413-arithmetic-slices/0413-arithmetic-slices.java
class Solution {
    public static int[] arr;

    public static int f_bu(){
        int[] dp = new int[5005];
        int ans = 0;
        for(int i = 0; i<arr.length; i++){
            if(i == 0 || i == 1) dp[i] = 0;
            else{
                if(arr[i] - arr[i-1] != arr[i-1] - arr[i-2]){
                    dp[i] = 0;
                }else{
                    dp[i] = 1 + dp[i-1];
                }
            }
        }

        for(int i = 0; i<arr.length; i++){
            ans+= dp[i];
        }

        return ans;
    }

    public int numberOfArithmeticSlices(int[] nums) {
        arr = nums;
        return f_bu();
    }
}